### ISSUE TITLE
DropModel event should return dropped model instance and index in dragulaModel

### ISSUE THREAD

**Issue by Soviut** [valor-software/ng2-dragula#306]
I think the DropModel event is the perfect opportunity to include extra event args including the instance that was dropped and the index of that instance within the parent model array. Essentially encapsulating the following.

```javascript
    private onDrop(args) {
        let [el, parent] = args;

        let index = this.getElementIndex(el);
        let item = this.items[index];
    }
```

**Comment by BojanKogoj** [comment]
It should be easy to get index, but this is the only way I found (source: https://github.com/bevacqua/dragula/issues/20)
```
private getElementIndex(el: any) {
    return [].slice.call(el.parentElement.children).indexOf(el);
}
```

**Comment by katesowles** [comment]
A decent Github thread on the topic : http://stackoverflow.com/questions/41007356/angular-2-dragula-model-updating-incorrectly/43083873#43083873

**Comment by cmddavid** [comment]
I used this approach:

```
        let that = this;
        dragulaService.dropModel.subscribe((args: any) => {
            // this seems to trigger all bags, but no solution available yet: https://github.com/valor-software/ng2-dragula/issues/558
            let [bagName, el, target, source] = args;
            if(this.bag.nativeElement === target){ // needed since dropModel triggers for all bags, not only on the dropped bag
                let transferData:any = {
                    taskId: Number(el.dataset.id),
                    oldLabelId: Number(source.dataset.id),
                    newLabelId: Number(target.dataset.id),
                    oldIndex: Number(el.dataset.index)
                }
                for(let i = 0; i < that.tasks.length; i++){
                    let task = that.tasks[i];
                    if(task.id == transferData.taskId){
                        transferData.newIndex = i;
                        console.log(transferData);
                        break;
                    }
                }
            }
        });
```

**Comment by louisamoros** [comment]
@cmddavid could you make a plnkr? I like your solution but I can't make it work with the `[attr.data-id]` attribute.

**Comment by cmddavid** [comment]
@louisamoros, sorry for the late reply. Are you still in need? I can make you one tonight if so.

**Comment by meelan** [comment]
@cmddavid , yes please make the plunkr

**Comment by cmddavid** [comment]
@louisamoros @meelan 

Here it is, but to save some time and complexity it only works when dropping items from bag2 to bag1.

https://plnkr.co/edit/S55TLdOcq5Z0YQRGiMxo?p=preview

I hope you guys get the idea, if you need the complete example working the other way around as well i will look into it in the weekend.

**Comment by alexandreannic** [comment]
To retrieve the new index in case you have multiples bags, you can do something like that :

```
constructor(private dragulaService: DragulaService) {
    this.dragulaService.drop.subscribe((args: any) => {
        const [bagName, elSource, bagTarget, bagSource, elTarget] = args;
        const newIndex = elTarget ? getElementIndex(elTarget) : bagTarget.childElementCount;
    })
}

private getElementIndex(el: HTMLElement): number {
    return [].slice.call(el.parentElement.children).indexOf(el);
}
```

It's a bit tricky because `elTarget` is the element where the dragged one is dropped. Therefore, if we drop at the end of another bag, that element is null. In this case we can retrieve the number of elements from the targeted bag.

**Comment by AhHa45** [comment]
The whole library is useless for server interaction if you can't get the new/old index of the dropped item...

**Comment by AhHa45** [comment]
alright in case for ngFor I found a simple solution:

add dragulaModel to your bag
(https://github.com/valor-software/ng2-dragula#dragulamodel)
this synchronizes the index of your array with the index after the drop event.
A detailed example:
https://stackblitz.com/edit/angular-3f61gg?file=app%2Fapp.component.html

**Comment by danielsmeyer** [comment]
@cormacrelf By done in v2 do you mean there is a method to return the old and new index? I can't seem to find that.

**Comment by cormacrelf** [comment]
@danielsmeyer It’s all over the docs. Events are better and dropModel and removeModel get more properties.

**Comment by angelika** [comment]
@danielsmeyer Did you ever find it? Cause I sure can't... I get the model instances and all that, but not the old and new index.

**Comment by cormacrelf** [comment]
Just use Array.prototype.findIndex.

**Comment by angelika** [comment]
@cormacrelf Yeah, thanks. I know that. Just got the impression when you said it was "done", that the information would be available. My bad, I guess.

**Comment by cormacrelf** [comment]
Can I ask what is the use case?

**Comment by cormacrelf** [comment]
If people really want indices, you can have them, I guess.

**Comment by angelika** [comment]
In my case I use it to save the new sort order of the items in a list to the server, which I would assume is quite common behaviour. My objects are quite complex and it is not practical to send the entire list of objects to the server, so I send only the id of the object and the it's new sort order. 

I'm grateful for this module and if/how you choose to implement features is up to you, but I do think it would be a very welcomed addition.

**Comment by cormacrelf** [comment]
There’s also `newArray.map(x => x.id)`?

**Comment by cormacrelf** [comment]
Personally I wouldn’t send JUST the moved object’s ID and new index, because that’s a recipe for inconsistency. Requests will fail, some will arrive out of order, and you’ll get objects with the same list position or strange non-linearizable results eventually.

Mostly because the “from” position is implicit in state as seen by a single save call. And what you’re really trying to express is “a whole bunch of items just got shifted around”, not just one. Two solutions spring to mind:

- send the complete ordering, or any changes each time. Doesn’t cost much if you just send IDs. Conceptually you can compute “just the changes” by zipping the old and new arrays together into the tuple (index, idOld, idNew), and pulling out any results where the IDs are different.
- use Trello-style floating point ordering with random jitter. More complex but extremely O(1).

But yeah, I’ll add indexes.

**Comment by cormacrelf** [comment]
v2.1.0 released
